### PR TITLE
Modernization + AZStd::function compare fix.

### PR DESCRIPTION
**Pull request by nemerle**
Modernize ( `bool`/`override`/other) code in AzCore, AzFramework, AzQtComponents, AzToolsFramework, etc.
Replaced a `bind` or two, use `using` in a few places as well.

Fix nullptr comparison of AZStd::function.

### CODE REVIEW

File: Code/Framework/AzCore/AzCore/Memory/AllocatorManager.cpp [hunk 1]
```diff
@@ -39,9 +39,9 @@ namespace AZ
                 return AZStd::hash<AZStd::string_view>{}(key);
             }
         };
-        typedef AZStd::basic_string<char, AZStd::char_traits<char>, AZStdIAllocator> AMString;
-        typedef AZStd::unordered_map<AMString, IAllocator*, AMStringHasher, AZStd::equal_to<>, AZStdIAllocator> AllocatorNameMap;
-        typedef AZStd::unordered_map<AMString, AMString, AMStringHasher, AZStd::equal_to<>, AZStdIAllocator> AllocatorRemappings;
+        using AMString = AZStd::basic_string<char, AZStd::char_traits<char>, AZStdIAllocator>;
+        using AllocatorNameMap = AZStd::unordered_map<AMString, IAllocator *, AMStringHasher, AZStd::equal_to<>, AZStdIAllocator>;
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzCore/AzCore/Memory/AllocatorManager.cpp:43]: nitpick: space between IAllocator and *
```suggestion
        using AllocatorNameMap = AZStd::unordered_map<AMString, IAllocator*, AMStringHasher, AZStd::equal_to<>, AZStdIAllocator>;
```

File: Code/Framework/AzCore/AzCore/Serialization/SerializeContext.cpp [hunk 2]
```diff
@@ -2182,18 +2182,18 @@ namespace AZ
 
     AZ::SerializeContext::DataPatchUpgradeHandler::~DataPatchUpgradeHandler()
     {
-        for (auto fieldUpgrades : m_upgrades)
+        for (const auto &fieldUpgrades : m_upgrades)
         {
-            for (auto versionUpgrades : fieldUpgrades.second)
+            for (const auto &versionUpgrades : fieldUpgrades.second)
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzCore/AzCore/Serialization/SerializeContext.cpp:2187]: coding standards: the ampersand should be on the auto, not the variable.

File: Code/Framework/AzCore/AzCore/Serialization/SerializeContext.h [hunk 3]
```diff
@@ -1938,7 +1938,7 @@ namespace AZ
         if (m_context->IsRemovingReflection())
         {
             // Delete any attributes allocated for this call.
-            for (auto attributePair : attributes)
+            for (auto &attributePair : attributes)
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzCore/AzCore/Serialization/SerializeContext.h:1941]: coding standards: ampersand should be on the auto

```suggestion
            for (auto& attributePair : attributes)
```

Also most of these range based for loops don't need to change.
They are iterating over pointers.

Comment (nemerle) [reply Code/Framework/AzCore/AzCore/Serialization/SerializeContext.h:1941]: Well, this specific one is iterating over pairs that contain the pointer :)

File: Code/Framework/AzCore/AzCore/std/function/function_base.h [hunk 4]
```diff
@@ -22,7 +22,7 @@
 #include <AzCore/std/createdestroy.h>
 
 #define AZSTD_FUNCTION_TARGET_FIX(x)
-#define AZSTD_FUNCTION_ENABLE_IF_NOT_INTEGRAL(Functor, Type)  AZStd::enable_if_t<!std::is_integral_v<Functor>, Type>
+#define AZSTD_FUNCTION_ENABLE_IF_NOT_INTEGRAL(Functor, Type)  AZStd::enable_if_t<!std::is_integral_v<Functor>&&!std::is_null_pointer_v<Functor>, Type>
```
Comment (lumberyard-employee-dm) [inline Code/Framework/AzCore/AzCore/std/function/function_base.h:25]: spacing: lack of spacing around the logical-and
```suggestion
#define AZSTD_FUNCTION_ENABLE_IF_NOT_INTEGRAL(Functor, Type)  AZStd::enable_if_t<!std::is_integral_v<Functor> && !std::is_null_pointer_v<Functor>, Type>
```

File: Code/Framework/AzCore/AzCore/Serialization/SerializeContext.h (hunk 3 above)
Comment (lumberyard-employee-dm) [reply Code/Framework/AzCore/AzCore/Serialization/SerializeContext.h:1941]: You are right, my mistake.

**Review by lumberyard-employee-dm — APPROVED** [review]

**Review by AMZN-ScottR — APPROVED** [review]

**Pull request merged**